api/tokens.go

```diff
@@ -148,3 +151,72 @@ func (api *API) RefreshHandler(ctx context.Context, w http.ResponseWriter, req *
 
 	return models.NewSuccessResponse(jsonResponse, http.StatusCreated, headers), nil
 }
+
+//SignOutAllUsersHandler bulk refresh token invalidation for panic sign out handling
+func (api *API) SignOutAllUsersHandler(ctx context.Context, w http.ResponseWriter, req *http.Request) (*models.SuccessResponse, *models.ErrorResponse) {
+	usersList := models.UsersList{}
+	listUserInput := usersList.BuildListUserRequest("status=\"Enabled\"", "username", int64(0), &api.UserPoolId)
+	listUserResp, err := api.CognitoClient.ListUsers(listUserInput)
+	if err != nil {
+		return nil, models.NewErrorResponse(http.StatusInternalServerError, nil, err)
+	}
+	usersList.MapCognitoUsers(listUserResp)
+	globalSignOut := &models.GlobalSignOut{
+		ResultsChannel: make(chan string, len(usersList.Users)),
+		BackoffSchedule: []time.Duration{
+			1 * time.Second,
+			3 * time.Second,
+			10 * time.Second,
+		},
+		RetryAllowed: true,
+	}
+	// run api.SignOutUsersWorker concurrently
+	go api.SignOutUsersWorker(ctx, globalSignOut, &usersList.Users)
+	return models.NewSuccessResponse(nil, http.StatusAccepted, nil), nil
+}
+
+// SignOutUsersWorker - signs out users globally by invalidating user's refresh token
+func (api *API) SignOutUsersWorker(ctx context.Context, g *models.GlobalSignOut, usersList *[]models.UserParams) {
```

Comment:
Observations from me.
- We are calling this method from a Go routine so the API request is async. But we are iterating through the users within that go routine in series. I'm unaware of the requirements discussions you've already had so just wondering if this was a requirement or decision made intentionally?

   If not (at this risk of making this even more complex) I am wondering if we can/should make this process concurrent as well? i.e. process the list of users concurrently within this go routine?

- This is a complex block of code and there is a lot going on here. I think we should consider breaking this down a bit into smaller methods to make it a bit easier to digest.

I appreciate this is a complex problem with very restrictive requirements Mike and you've invested time and effort into working out how to solve it. Im also unfamiliar with the nuances of it so forgive me for asking any dumb questions.

Reply:
Thanks Dave. Yes, good point. I discussed with James. The AWS Cognito `AdminUsersGlobalSignOut` API restricts requests to 25 requests/second.

We felt that distributing the workload to `n` number of workers would be negated by the this request limit. Thats why I've implemented the exponential backoff feature as this single concurrent worker will hit that and will have to backoff to ensure that it adheres to this limit too to avoid DDOS and possible permanent blacklisting.